### PR TITLE
[4/6][kimi-deterministic] Fix sampling positions under DP-attention MLP-sync padding

### PR DESCRIPTION
**Pull request by zyzshishui**
> **Stacked series (4/6)**: stacked on `kimi-deterministic/3-deepep-normal-bf16-deepgemm` (#30819), so the diff below includes 1-3/6. This PR's own changes are the last 3 commits (ending b8f498fcf — the last clears the new MLP-sync field on TBO child batches). Combined CI: #30822.

## Motivation

With DP attention, seeded sampling crashes ("Seed and positions must have the same shape"): MLP-sync padding extends `positions`/`seq_lens` while `SamplingBatchInfo` covers only real requests. Repro: `--enable-dp-attention --dp-size 2`, /generate with `temperature > 0` + `sampling_seed`. (Part 4/6 of the kimi-deterministic series.)

## Modifications

Fix at the root instead of at the consumer (per review): `prepare_mlp_sync_batch` records the pre-padding token count, and the non-speculative arm of `post_forward_mlp_sync_batch` now slices `positions`/`seq_lens`/`req_pool_indices`/`seq_lens_cpu` back to the real rows, mirroring the existing draft-decode unpadding. This keeps every post-forward consumer (seeded sampling, ngram token-table updates, prefill-only dummy ids) free of padding; the earlier `ModelRunner.sample`-side slice is removed as dead. A unit test locks the pad/unpad round-trip (`test/registered/unit/model_executor/test_mlp_sync_pad_unpad.py`).

## Accuracy Tests

Repro fixed (dp2: crash before, samples after; e2e campaign below ran with the original sample-side slice — the reworked fix feeds the sampler the same rows, and the new unit test covers the round-trip directly). Kimi-K2-Instruct-BF16, 16 GPUs (TP16 / DP8-attention / EP16, fa4 + deep_gemm + deepep normal, radix on), 128 prompts (1k-16k) x 512 tokens, abs tol 0.0: replay, shuffled concurrency-32, bs1 vs bs128, cached-prefix extend vs cold prefill, decode logprobs vs prefill rescoring (65,536 tokens) — all bitwise identical.

## Speed Tests and Profiling

No impact (tensor slices in the once-per-forward post-sync path).

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: [Run #29979964703](https://github.com/sgl-project/sglang/actions/runs/29979964703)
Latest PR Test (Extra): :x: [Run #29979964525](https://github.com/sgl-project/sglang/actions/runs/29979964525)

### CODE REVIEW

File: python/sglang/srt/model_executor/model_runner.py [hunk 1]
```diff
@@ -3239,19 +3239,27 @@ def sample(
...
+        # sampling_info only covers the real requests, so slice back to the
+        # sampled rows (real rows come first; padding is appended).
+        positions = (
+            forward_batch.positions
+            if forward_batch.forward_mode.is_decode()
+            else forward_batch.seq_lens - 1
+        )
+        num_sampled = forward_batch.sampling_info.temperatures.shape[0]
+        if positions.shape[0] != num_sampled:
+            positions = positions[:num_sampled]
```
Comment (ch-wan) [inline python/sglang/srt/model_executor/model_runner.py]: **[suggestion]** The mismatch is created in `ForwardBatch.prepare_mlp_sync_batch` / `_pad_inputs_to_size` and only partially cleaned up in `post_forward_mlp_sync_batch`. Non-speculative post-forward restores `batch_size` and trims logits, but leaves `positions`/`seq_lens`/`req_pool_indices` padded; the draft-decode branch already unpads `positions` and `seq_lens`. Fixing only at `ModelRunner.sample` works for this crash, but other post-forward consumers (e.g. `maybe_update_ngram_token_table` assigning `seq_lens` into a `[:batch_size]` slice, or prefill-only dummy ids sized with `len(seq_lens)`) can still see padded tensors.

**Suggestion:** Prefer unpadding `positions`/`seq_lens` (and related per-req tensors) to `_original_batch_size` inside non-spec `post_forward_mlp_sync_batch`, mirroring the draft-decode path; keep this sample-side slice as a defensive belt if desired.

File: python/sglang/srt/model_executor/model_runner.py [hunk 2]
```diff
@@ -3239,19 +3239,27 @@ def sample(
...
+        # For prefill, we only use the position of the last token. DP-attention
+        # MLP-sync padding appends dummy rows to positions/seq_lens while
+        # sampling_info only covers the real requests, so slice back to the
+        # sampled rows (real rows come first; padding is appended).
+        positions = (
+            forward_batch.positions
+            if forward_batch.forward_mode.is_decode()
+            else forward_batch.seq_lens - 1
+        )
+        num_sampled = forward_batch.sampling_info.temperatures.shape[0]
```
Comment (ch-wan) [inline python/sglang/srt/model_executor/model_runner.py]: **[nit]** `forward_batch.sampling_info.temperatures.shape[0]` is correct (`temperatures` is shaped `[num_reqs, 1]` and matches `sampling_seed` / logits rows), but it is an indirect way to spell "number of sampled rows". After post-forward, `forward_batch.batch_size` is restored to the real request count, and `len(forward_batch.sampling_info)` is defined as `len(self.temperatures)`.

**Suggestion:** Prefer `len(forward_batch.sampling_info)` or `forward_batch.batch_size` for readability, unless there is a known case where temperatures diverge from restored batch size (none found for this path).

File: python/sglang/srt/model_executor/model_runner.py [hunk 3]
```diff
@@ -3239,19 +3239,27 @@ def sample(
...
+        if positions.shape[0] != num_sampled:
```
Comment (ch-wan) [inline python/sglang/srt/model_executor/model_runner.py]: **[suggestion]** The branch only handles `positions.shape[0] != num_sampled` by prefix-slicing. That is right when MLP-sync appends padding (verified: `_pad_tensor_to_size` always concatenates dummy rows after real data). If positions were ever shorter than `num_sampled`, the slice is a no-op and the original murmur assert still fires with little diagnosis. No unit test is included in this PR to lock the DP-attention + `sampling_seed` contract.

**Suggestion:** Optionally `assert positions.shape[0] >= num_sampled` before slicing. Add a small unit/integration test that builds a padded `positions`/`seq_lens` with real rows first and asserts seeded sample succeeds and uses only the first `num_sampled` positions (repro from the PR description: dp-attention, temperature > 0, sampling_seed).
